File: ruby/lib/xstream/x25519hkdf.rb
# encoding: binary
# frozen_string_literal: true

module XStream
  # XSTREAM with X25519 key agreement and HKDF as the key derivation function
  module X25519HKDF
    # Domain separation string passed as HKDF info
    HKDF_INFO = "XSTREAM_X25519_HKDF".freeze

    # Size of an AES-128 key * 2 (for SIV mode)
    SYMMETRIC_KEY_SIZE = 32

    # XSTREAM encryptor with X25519+HKDF key derivation
    class Encryptor < ::Miscreant::STREAM::Encryptor
      # Generate an XSTREAM encryptor object with a random ephemeral key
      #
      # @param public_key [String] 32-byte X25519 public key (i.e. compressed Montgomery-u coordinate)
      # @param encryption_alg [String] symmetric encryption algorithm to use with STREAM (default `"AES-PMAC-SIV"`)
      # @param digest_alg [String] digest algorithm to use with HKDF (default `"SHA256"`)
      # @param salt [String] (optional) salt value to pass to HKDF
      # @param csrng [#random_bytes] secure RNG to use to derive ephemeral X25519 key (default `SecureRandom`)
      #
      # @return [Array(XStream::X25519HKDF, String)] STREAM encryptor and ephemeral public key
      def self.generate(
        public_key,
        encryption_alg: "AES-PMAC-SIV",
        digest_alg: "SHA-256",
        salt: nil,
        csrng: SecureRandom
      )
        ephemeral_scalar = csrng.random_bytes(::X25519::KEY_SIZE)
        ephemeral_public = ::X25519.calculate_public_key(ephemeral_scalar)

        symmetric_key = X25519HKDF.kdf(
          ephemeral_scalar,
          public_key,
          salt: salt,
          digest_alg: digest_alg,
          output_size: SYMMETRIC_KEY_SIZE
        )

        stream = new(encryption_alg, symmetric_key, XStream::NONCE)
        [stream, ephemeral_public]
      end
    end

    # XSTREAM decryptor class with X25519+HKDF key derivation
    class Decryptor < ::Miscreant::STREAM::Decryptor
      # Create an XSTREAM decryptor object using our private key and an ephemeral public key
      #
      # @param private_key [String] 32-byte X25519 private key (i.e. private scalar)
      # @param ephemeral_public [String] 32-byte X25519 ephemeral public key from XSTREAM encryption
      # @param encryption_alg [String] symmetric encryption algorithm to use with STREAM (default `"AES-PMAC-SIV"`)
      # @param digest_alg [String] digest algorithm to use with HKDF (default `"SHA256"`)
      # @param salt [String] (optional) salt value to pass to HKDF
      def initialize(
          private_key,
          ephemeral_public,
          encryption_alg: "AES-PMAC-SIV",
          digest_alg: "SHA-256",
          salt: nil
      )
        # Perform an X25519 elliptic curve Diffie-Hellman operation and use
        # the resulting shared secret to derive a symmetric key (using HKDF)
        symmetric_key = X25519HKDF.kdf(
          private_key,
          ephemeral_public,
          salt: salt,
          digest_alg: digest_alg,
          output_size: SYMMETRIC_KEY_SIZE
        )

        super(encryption_alg, symmetric_key, XStream::NONCE)
      end
    end

    # Derive a symmetric encryption key from the combination of a public and
    # private key and salt using X25519 D-H and HKDF
    def self.kdf(private_key, public_key, output_size:, salt: nil, digest_alg: "SHA-256")
      raise ArgumentError, "invalid digest_alg: #{digest_alg}" unless digest_alg == "SHA-256"

      # Use X25519 to compute a shared secret
      shared_secret = X25519.diffie_hellman(private_key, public_key)

      # Use HKDF to derive a symmetric encryption key from the shared secret
      ::HKDF.new(
        shared_secret,
        salt: salt,
        info: HKDF_INFO,
        algorithm: "SHA256"
      ).next_bytes(output_size)
    end
  end
end

File: python/xstream/x25519hkdf.py
"""x25519hkdf.py: Implementation of XSTREAM's core cryptographic primitive combining X25519+HKDF+STREAM"""

import os
from cryptography.hazmat.primitives.asymmetric.x25519 import (X25519PrivateKey, X25519PublicKey)
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.kdf.hkdf import HKDF
from cryptography.hazmat.backends import default_backend

from miscreant import stream

# Domain separation string passed as HKDF info
HKDF_INFO = b"XSTREAM_X25519_HKDF"

# Size of an X25519 key
X25519_KEY_SIZE = 32

# Size of an AES-128 key * 2 (for SIV mode)
SYMMETRIC_KEY_SIZE = 32

# STREAM nonce of all zeroes (since we always derive a unique key per STREAM)
NONCE = b"\0\0\0\0\0\0\0\0"


class Encryptor(stream.Encryptor):
    """XSTREAM encryptor with X25519+HKDF key derivation"""

    @staticmethod
    def generate(
        public_key,
        encryption_alg="AES-PMAC-SIV",
        digest_alg=hashes.SHA256(),
        salt=None,
        csrng=os.urandom,
    ):
        """
        Generate an XSTREAM encryptor object with a random ephemeral key

        :param public_key: 32-byte X25519 public key (i.e. compressed Montgomery-u coordinate)
        :param encryption_alg: symmetric encryption algorithm to use with STREAM (default "AES-PMAC-SIV")
        :param digest_alg: digest algorithm to use with HKDF (default "SHA256")
        :param salt: (optional) salt value to pass to HKDF (default None)
        :param csrng: (optional) secure random number generator used to generate ephemeral key (default os.urandom)
        :return: STREAM encryptor and ephemeral public key
        """
        ephemeral_scalar = X25519PrivateKey._from_private_bytes(csrng(X25519_KEY_SIZE))

        symmetric_key = kdf(
          private_key=ephemeral_scalar,
          public_key=X25519PublicKey.from_public_bytes(public_key),
          digest_alg=digest_alg,
          length=SYMMETRIC_KEY_SIZE,
          salt=salt
        )

        enc = Encryptor(encryption_alg, symmetric_key, NONCE)
        return enc, ephemeral_scalar.public_key().public_bytes()


class Decryptor(stream.Decryptor):
    """XSTREAM decryptor class with X25519+HKDF key derivation"""

    def __init__(
        self,
        private_key,
        ephemeral_public,
        encryption_alg="AES-PMAC-SIV",
        digest_alg=hashes.SHA256(),
        salt=None
    ):
        """
        Create an XSTREAM decryptor object using our private key and an ephemeral public key

        :param private_key: 32-byte X25519 private key (i.e. private scalar)
        :param ephemeral_public: 32-byte X25519 ephemeral public key from XSTREAM encryption
        :param encryption_alg: symmetric encryption algorithm to use with STREAM (default "AES-PMAC-SIV")
        :param digest_alg: digest algorithm to use with HKDF (default "SHA256")
        :param salt: (optional) salt value to pass to HKDF (default None)
        """

        # Perform an X25519 elliptic curve Diffie-Hellman operation and use
        # the resulting shared secret to derive a symmetric key (using HKDF)
        symmetric_key = kdf(
          private_key=X25519PrivateKey._from_private_bytes(private_key),
          public_key=X25519PublicKey.from_public_bytes(ephemeral_public),
          digest_alg=digest_alg,
          length=SYMMETRIC_KEY_SIZE,
          salt=salt
        )

        super(Decryptor, self).__init__(encryption_alg, symmetric_key, NONCE)


def kdf(
    private_key,
    public_key,
    length,
    salt=None,
    digest_alg=hashes.SHA256(),
    backend=default_backend()
):
    """
    Derive a symmetric encryption key from the combination of a public and
    private key and salt using X25519 D-H and HKDF
    """
    # Use HKDF to derive a symmetric encryption key from the shared secret
    hkdf = HKDF(
        algorithm=digest_alg,
        length=length,
        salt=salt,
        info=HKDF_INFO,
        backend=backend
    )

    # Use X25519 to compute a shared secret
    return hkdf.derive(private_key.exchange(public_key))
